### PR TITLE
lxd/storage/drivers/zfs: Use -w flag with `zfs send` if possible

### PR DESCRIPTION
**Pull request by monstermunchkin**
- lxd/storage/drivers/zfs: Check for encryption
- lxd/storage/drivers/zfs: Handle encrypted send

### CODE REVIEW

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 1]
```diff
@@ -548,7 +548,19 @@ func (d *zfs) CreateVolumeFromCopy(vol Volume, srcVol Volume, copySnapshots bool
 
 		// Handle transferring snapshots.
 		if len(snapshots) > 0 {
-			sender = exec.Command("zfs", "send", "-R", srcSnapshot)
+			args := []string{"send", "-R"}
+
+			// Check if dataset is encrypted. If so, add -w.
```
[the remark this answers is not in the document]
Comment (simondeziel) [reply lxd/storage/drivers/driver_zfs_volumes.go:553]: Agreed, `-w` degrading to `-Lec` when no encryption is used has nice benefits (basically skips the decompression/re-compression) when you copy FSes around.

[the remark this answers is not in the document]
Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs_volumes.go]: Done

File: lxd/storage/drivers/driver_zfs.go [hunk 2]
```diff
@@ -22,6 +22,7 @@ var zfsVersion string
 var zfsLoaded bool
 var zfsDirectIO bool
 var zfsTrim bool
+var zfsEncryption bool
```
[the remark this answers is not in the document]
Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs.go]: Done

**Review by simondeziel — APPROVED** [review]

**Comment by tomponline** [comment]
It would be nice to explain why raw mode is preferable if supported.

**Comment by stgraber** [comment]
Not so much preferable as required if dealing with encrypted datasets.

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 3]
```diff
@@ -548,7 +548,16 @@ func (d *zfs) CreateVolumeFromCopy(vol Volume, srcVol Volume, copySnapshots bool
 
 		// Handle transferring snapshots.
 		if len(snapshots) > 0 {
-			sender = exec.Command("zfs", "send", "-R", srcSnapshot)
+			args := []string{"send", "-R"}
+
+			// Use raw flag if possible.
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_volumes.go:553]: Can you make this `Use raw flag is supported, this is required to send/receive encrypted volumes (and enables compression).`

Comment (monstermunchkin) [reply lxd/storage/drivers/driver_zfs_volumes.go]: Done

**Comment by simondeziel** [comment]
> It would be nice to explain why raw mode is preferable if supported.

Sending in raw mode avoids decompressing the dataset that is streamed. This is faster and results in smaller streams if the dataset was compressed to begin with.

For encrypted datasets, using raw mode has 2 advantages: you don't need to load the key and you are not sending plaintext datasets. This is useful when backing up to a untrusted server.

**Comment by tomponline** [comment]
> > It would be nice to explain why raw mode is preferable if supported.
> 
> Sending in raw mode avoids decompressing the dataset that is streamed. This is faster and results in smaller streams if the dataset was compressed to begin with.
> 
> For encrypted datasets, using raw mode has 2 advantages: you don't need to load the key and you are not sending plaintext datasets. This is useful when backing up to a untrusted server.

thanks!

**Pull request merged**